svm/src/transaction_processor.rs

```diff
@@ -620,14 +624,16 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
             if let Some((key, program_to_recompile)) = program_cache.programs_to_recompile.pop() {
                 let effective_epoch = program_cache.latest_root_epoch.saturating_add(1);
                 drop(program_cache);
+                let program_cache_read = self.program_cache.read().unwrap();
                 if let Some(recompiled) = load_program_with_pubkey(
                     callbacks,
-                    &self.get_environments_for_epoch(effective_epoch),
+                    &program_cache_read.get_environments_for_epoch(effective_epoch),
                     &key,
                     self.slot,
                     &self.epoch_schedule,
                     false,
                 ) {
+                    drop(program_cache_read);
```

Comment:
If we don't go with my suggestion above about getting `environments_for_epoch()` with the write lock, I think we could obtain it from a read lock in a 1-liner to avoid the need for this explicit `drop`:
```rust
let environments_for_epoch = self.program_cache.read().unwrap().get_environments_for_epoch(effective_epoch);
```

Reply:
Yes, this will work too. I am just trying to restore this part of the code to the original version as much possible.
This can be done as a follow on PR to refactor this out.

Reply:
Gotcha, so basically same sentiment as other item then. Works for me, and we can evaluate further changes such as what I proposed elsewhere